tensorflow/compiler/xla/service/gpu/conv_layout_normalization.cc

```diff
@@ -105,14 +105,17 @@ StatusOr<HloInstruction*> UpdateLayoutForCudnnConvolution(
 
   Shape normalized_shape;
   if (hlo->shape().IsTuple()) {
-    TF_RET_CHECK(hlo->shape().tuple_shapes_size() == 2);
-    TF_RET_CHECK(hlo->shape().tuple_shapes(1).rank() == 1)
-        << "Second element in a convolution tuple is expected to be an "
+    TF_RET_CHECK(hlo->shape().tuple_shapes().back().rank() == 1)
```

Comment:
I would check that all but the last and first elements have rank zero. Then you can get rid of some of the for-loops you added that change the layout of all the outputs

Admittedly, I am having trouble understanding this file and it would take me a long time to understand it and verify the changes you made, especially on a hypothetical non-scalar second output. Still, I think only changing the layout of the first element makes things simpler.

Reply:
When we further generalize the graph convolutions, the result tensor may no longer be the first element of the tuple. The workspace will remain the final element though.